tools/export/sw4stm32/__init__.py

```diff
@@ -14,97 +14,357 @@
+        str must be converted with win_to_unix() before using this function.
+        """
+        if path is None:
+            return None
+        if path[:2] == './':
+            return path[2:]
+        return path
+
+    def build_excludelist(self):
+        """
+        This method creates list for excluded directories.
+        """
+        self.source_folders = [self.filter_dot(s)
+           for s in set(dirname(src) for src in
+               self.resources.c_sources +
+               self.resources.cpp_sources +
+               self.resources.s_sources)]
+        if '.' in self.source_folders:
+            self.source_folders.remove('.')
+
+        top_folders = [f for f in set(s.split('/')[0]
+           for s in self.source_folders)]
+
+        for top_folder in top_folders:
+            for root, dirs, files in walk(top_folder, topdown=True):
+                # Paths returned by os.walk() must be split with os.dep
+                # to accomodate Windows weirdness.
+                parts = root.split(sep)
+                self.remove_unused('/'.join(parts))
+
+    def remove_unused(self, path):
+        """
+        Method for checking if path is needed.
+        Method adds path to excluded list if not needed
+        and is not subdirectory of already excluded directory
+        """
+        found = False
+        for used in self.include_path:
+            if path == used:
+                found = True
+        needtoadd = True
+        if not found:
+            for directory in self.exclude_dirs:
+               # Do not exclude subfolders from excluded folder
+                if path.find(directory+'/') != -1:
+                    needtoadd = False
+            if needtoadd:
+                self.exclude_dirs.append(path)
+
     def generate(self):
         fp_hardware = "no"
         fp_abi = "soft"
         core = self.toolchain.target.core
         if core == "Cortex-M4F" or core == "Cortex-M7F":
             fp_hardware = "fpv4-sp-d16"
-            fp_abi = "soft-fp"
+            fp_abi = "softfp"
         elif core == "Cortex-M7FD":
             fp_hardware = "fpv5-d16"
-            fp_abi = "soft-fp"
-            
+            fp_abi = "softfp"
+
+        self.resources.win_to_unix()
+
         libraries = []
         for lib in self.resources.libraries:
-            l, _ = splitext(basename(lib))
-            libraries.append(l[3:])
+            library, _ = splitext(basename(lib))
+            libraries.append(library[3:])
+
+        self.include_path = [self.filter_dot(s)
+           for s in self.resources.inc_dirs]
+        print 'Include folders: %d' % len(self.include_path)
+
+        self.exclude_dirs = []
+        self.build_excludelist()
+
+        print 'Exclude folders: %d' % len(self.exclude_dirs)
+
+        self.exclude_dirs = '|'.join(self.exclude_dirs)
+
+        ld_script = self.filter_dot(self.resources.linker_script)
+
+        # self.lib_dirs = [self.filter_dot(s) for s in self.resources.lib_dirs]
```

Comment:
Changes made. Can't promise anything about build method or other things, there was pretty tight timeline for this and sw4stm32 exporter was more broken and bare boned than we thought.